File: Data Structures & Algorithms/add-two-numbers/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        digit = 1
        num1 = 0
        while l1:
            num1 += l1.val*digit
            digit *= 10
            l1 = l1.next
        digit = 1
        num2 = 0
        while l2:
            num2 += l2.val*digit
            digit *= 10
            l2 = l2.next
        num = num1+num2
        head = ListNode(0)
        curr = head
        it = 0
        while num >= 0:
            value = num % 10
            num = num // 10
            curr.next = ListNode(value)
            curr = curr.next
            if num == 0:
                break
        return head.next

        

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        int count = 0;
        ListNode curr = head;
        while (curr!=null) {
            count += 1;
            curr = curr.next;
        }
        if (count == 0 || count < k) { // edge case if head is empty or less nodes than k
            return head;
        }
        //we can sort of slice the linked list into partitions, and keep track of the partitions
        // have a for loop while count // k, iterate through node
        Deque<ListNode> queue = new ArrayDeque<>();
        curr = head;
        ListNode prev = null;
        ListNode temp;
        int j = 0;
        for (int i = 0; i < count / k; ++i) {
            for ( j = 0; j < k; ++j) {
                temp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = temp;
            } // curr equal next head, prev will be new head, 1<-2<-3(prev) 4(curr) 5 6
            queue.offer(prev);
            prev = null;
        }
        if (curr!=null) {
            queue.offer(curr);
        }
        ListNode dummy = new ListNode(0);
        prev = dummy;
        while(!queue.isEmpty()) {
            curr = queue.pop();
            while (curr!= null) {
                prev.next = curr;
                prev = curr;
                curr = curr.next;
            }
        }
        return dummy.next;
    }
}
